Mathlib/Algebra/BigOperators/Finsupp.lean

```diff
@@ -583,3 +589,91 @@ theorem prod_pow_pos_of_zero_not_mem_support {f : ℕ →₀ ℕ} (nhf : 0 ∉ f
     pow_ne_zero _ fun H => by subst H; exact nhf hf
 
 end Nat
+
+namespace Submonoid
+
+variable {M : Type*} [CommMonoid M] {ι : Type*} (f : ι → M) (x : M)
+
+@[to_additive]
+theorem exists_finsupp_of_mem_closure_range (hx : x ∈ closure (Set.range f)) :
+    ∃ a : ι →₀ ℕ, x = a.prod (f · ^ ·) := by
+  classical
+  induction hx using closure_induction with
+  | mem x h => obtain ⟨i, rfl⟩ := h; exact ⟨Finsupp.single i 1, by simp⟩
+  | one => use 0; simp
+  | mul x y hx hy hx' hy' =>
+    obtain ⟨⟨v, rfl⟩, w, rfl⟩ := And.intro hx' hy'
+    use v + w
+    rw [Finsupp.prod_add_index]
+    · simp
+    · simp [pow_add]
+
+variable {f x} in
+@[to_additive]
+theorem mem_closure_range_iff :
+    x ∈ closure (Set.range f) ↔ ∃ a : ι →₀ ℕ, x = a.prod (f · ^ ·) := by
```

Comment:
I think linearCombination is defeq to Finsupp.sum, so the additive version of this can probability be golfed using [Finsupp.range_linearCombination](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Finsupp/LinearCombination.html#Finsupp.range_linearCombination). However as I said it requires too many imports and doesn't give you the multiplicative version.